README.md

```diff
@@ -1,28 +1,182 @@
+
+When using Velveteen in particular, try to follow:
+
+* Keep workers focused on a single task and interact with at most one service
+  (e.g. S3, Postgres, third-party API).
+* Name workers after the actions they perform. E.g. a worker that fetches a
+  user's most recent GitHub commits could be called `FetchUserCommits`.
+  Velveteen will use the same name for the worker's queue.
+* In general, routing keys should be past tense events. E.g. `FetchUserCommits`
+  could publish the API response with a routing key of `user.commits.fetched`.
+  This documents what already happened and makes no assumptions about what
+  happens next. If a worker is the first step in a pipeline, e.g. runs a
+  schedule, use present tense - `user.commits.fetch`.
+
+## Message validation
+
+Velveteen can validate incoming messages with a JSON Schema.
+
+```ruby
+Velveteen::Config.schema_directory = "app/schemas"
+
+class FetchUserCommits < Velveteen::Worker
+  ...
+  self.message_schema = "fetch_user_commits.json"
+
+  def perform
+    # do something with the GitHub API
+  end
+end
+```
+
+_app/schemas/fetch_user_commits.json_
+```json
+{
+  "type": "object",
+  "required": ["username"],
+  "properties": {
+    "username": {"type": "string"}
+  }
+}
+```
+
+When a message is added to this worker's queue and it does not match the schema,
+a `Velveteen::InvalidMessage` exception will be raised and handled by the
+configured error handler. The worker's `perform` method will not be invoked.
+
+## Rate limiting
+
+Limiting workers to processing a certain number of messages per minute is
+supported. This is implemented by periodically publishing tokens to a dedicated
+queue (with a max size of 1), where the worker must successfully take a token
+before it begins.
+
+To produce tokens, Velveteen needs the name of the queue to publish to and the
+number of messages per minute. For example, [GitHub's API limit] is 5000
+requests per hour (~83.3 per minute). With Velveteen, that looks like:
 
-To install the gem manually from your shell, run:
+[GitHub's API limit]: https://developer.github.com/v3/#rate-limiting
 
 ```shell
-gem install velveteen
+velveteen rate-limit github_tokens 83
 ```
 
-## Supported Ruby versions
+Limiting the worker can be enabled by specifying the rate limit queue:
+
+```ruby
+class FetchCommits < Velveteen::Worker
+  ...
+  self.rate_limit_queue = "github_tokens"
+
+  def perform
+    # fetch commits
+  end
+end
+```
 
-TODO: Figure this out
+Multiple workers can share the same token queue, which can be helpful when API
+calls to an external service are spread among many workers and are all governed
+by the same rate limit.
+
+The provided implementation does not support other techniques, such as bursting,
+throttling, or quotas.
+
+## Error handling
+
+### Message rejection
+
+The default error handler with log the error and reject the message without
+requeueing it.
+
+### Retry with exponential backoff
```

Comment:
As a reader, I immediately wondered if this is system-wide or per-worker.

Reply:
Technically it's system-wide, _but_ Velveteen only supports running a single worker at a time, so it could be configured differently "globally" for each worker.